internal/telemetry/metric_exporter.go

```diff
@@ -0,0 +1,341 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/aggregation"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+
+	"github.com/elastic/apm-data/model/modelpb"
+)
+
+// NewMetricExporter initializes a new MetricExporter
+func NewMetricExporter(p modelpb.BatchProcessor, opts ...ConfigOption) *MetricExporter {
+	cfg := newConfig(opts...)
+
+	return &MetricExporter{
+		processor: p,
+
+		metricFilter:        cfg.MetricFilter,
+		temporalitySelector: cfg.TemporalitySelector,
+		aggregationSelector: cfg.AggregationSelector,
+	}
+}
+
+// MetricExporter is an OpenTelemetry metric Reader which retrieves metrics,
+// filters them and emits them to the specified consumer
+type MetricExporter struct {
+	processor modelpb.BatchProcessor
+
+	metricFilter        []string
+	temporalitySelector metric.TemporalitySelector
+	aggregationSelector metric.AggregationSelector
+}
+
+// Temporality returns the Temporality to use for an instrument kind.
+func (e *MetricExporter) Temporality(k metric.InstrumentKind) metricdata.Temporality {
+	return e.temporalitySelector(k)
+}
+
+// Aggregation returns the Aggregation to use for an instrument kind.
+func (e *MetricExporter) Aggregation(k metric.InstrumentKind) aggregation.Aggregation {
+	return e.aggregationSelector(k)
+}
+
+func (e *MetricExporter) Export(ctx context.Context, rm *metricdata.ResourceMetrics) error {
+	batch := modelpb.Batch{}
+	now := time.Now()
+
+	baseEvent := modelpb.APMEvent{
+		Service: &modelpb.Service{
+			Name:     "apm-server",
+			Language: &modelpb.Language{Name: "go"},
+		},
+		Agent: &modelpb.Agent{
+			Name:    "internal",
+			Version: "unknown",
+		},
+		Event: &modelpb.Event{
+			Received: modelpb.FromTime(now),
+		},
+	}
+
+	for _, scopeMetrics := range rm.ScopeMetrics {
+		ms := make(metricsets)
+		for _, sm := range scopeMetrics.Metrics {
+			if e.isMetricFiltered(sm.Name) {
+				continue
+			}
+
+			if err := addMetric(sm, ms); err != nil {
+				return err
+			}
+		}
+
+		for key, ms := range ms {
+			event := baseEvent.CloneVT()
+			event.Timestamp = modelpb.FromTime(key.timestamp)
+			metrs := make([]*modelpb.MetricsetSample, 0, len(ms.samples))
+			for _, s := range ms.samples {
+				metrs = append(metrs, s)
+			}
+			event.Metricset = &modelpb.Metricset{Samples: metrs, Name: "app"}
+			if ms.attributes.Len() > 0 {
+				event.Labels = modelpb.Labels(event.Labels).Clone()
+				event.NumericLabels = modelpb.NumericLabels(event.NumericLabels).Clone()
```

Comment:
We may not need to clone them, I did it that way here because that's what apm-data does.
https://github.com/elastic/apm-data/blob/main/input/otlp/metadata.go#L322

Reply:
Actually, it seems the `Clone()` is used to initialize. Without it, I'm getting `panic: assignment to entry in nil map` panics.